lib/core/src/Cardano/Wallet/Api/Server.hs

```diff
@@ -2213,23 +2213,37 @@ balanceTransaction
     -> ApiBalanceTransactionPostData n
     -> Handler (ApiConstructTransaction n)
 balanceTransaction ctx genChange (ApiT wid) body = do
+    -- TODO: Most of the body of this function should really belong to
+    -- Cardano.Wallet to keep the Api.Server moduel free of business logic!
     pp <- liftIO $ NW.currentProtocolParameters nl
 
     --TODO  deal with coll and validity and delegations
-    let (Tx _id _fee _coll _inps outs wdrlMap mdM _validity) = txIncoming
-
-    ((FeeEstimation feeMin _), unsignedtx) <- withWorkerCtx ctx wid liftE liftE $ \wrk -> do
+    let (Tx _id _fee _coll _inps outs wdrlMap txMetadata _validity) = txIncoming
 
+    (delta, unsignedtx) <- withWorkerCtx ctx wid liftE liftE $ \wrk -> do
         (acct, _, _) <- liftHandler $ W.readRewardAccount @_ @s @k @n wrk wid
+
+        -- FIXME: This is wrong as it leads the underlying selection to
+        -- "believe" there's only one withdrawal and will be completely off
+        -- regarding the actual fee. See ADP-1182
         (wdrl, _) <- mkRewardAccountBuilder @_ @s @_ @n ctx wid $
             if Map.member acct wdrlMap
             then Just SelfWithdrawal
             else Nothing
-        let executionFee = calcScriptExecutionCost tl pp sealedTxIncoming
+
+        -- NOTE: It is not possible to know the script execution cost in
+        -- advance because it actually depends on the final transaction. Inputs
+        -- selected as part of the fee balancing might have an influence on the
+        -- execution cost.
+        -- However, they are bounded so it is possible to balance the
+        -- transaction considering only the maximum cost, and only after, try to
+        -- adjust the change and ExUnits of each redeemer to something more
+        -- sensible than the max execution cost.
+        let maxCost = maxScriptExecutionCost tl pp sealedTxIncoming
```

Comment:
It just occurred to me: `actualCost <= maxCost` only holds if `pp` is the protocol parameters active when the tx is accepted, i.e. if `pp` and the protocol parameter hash match.

Not sure if we, or the client will set the protocol parameter hash, but we should make sure it always matches with `pp`.

Reply:
The protocol parameter hash is only needed for validators who depends on it. I'd expect the PAB (or really, anyone constructing the transaction) to set it to what is expected.